apps/emqx_s3/src/emqx_s3_schema.erl

```diff
@@ -169,6 +170,27 @@ desc(s3_upload) ->
 desc(transport_options) ->
     "Options for the HTTP transport layer used by the S3 client".
 
+validations() ->
```

Comment:
Since the values are not cross roots, so there is no maybe need to add a root level validator which is evaluated by all config checks even if it's not checking this root.
IIRC, we can implement roots/0 like this.

```
roots() ->
  [{s3, hoconsc:mk(hoconsc:ref(s3), #{validator => fun(Conf) -> ... end})].
```

Reply:
What do you mean by "cross roots"?

Reply:
I mean if one needs to verify `root1.key1 = bar` is valid given `root0.key0 = foo`, then it's a cross-root validation,
one would have to add a global validator.

e.g. the recently added `cluster.discoverty_strategy = dns` vs `node.name = emqx@FQDN` check.